File: app/src/main/java/com/elementary/tasks/core/view_models/day_view/DayViewViewModel.java
package com.elementary.tasks.core.view_models.day_view;

import android.app.Application;
import android.widget.Toast;

import com.elementary.tasks.R;
import com.elementary.tasks.birthdays.DayViewProvider;
import com.elementary.tasks.birthdays.EventsDataSingleton;
import com.elementary.tasks.birthdays.EventsItem;
import com.elementary.tasks.birthdays.EventsPagerItem;
import com.elementary.tasks.birthdays.work.DeleteBirthdayFilesAsync;
import com.elementary.tasks.core.controller.EventControlFactory;
import com.elementary.tasks.core.data.models.Birthday;
import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.view_models.BaseDbViewModel;
import com.elementary.tasks.core.view_models.Commands;
import com.elementary.tasks.reminder.work.UpdateFilesAsync;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;

/**
 * Copyright 2018 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class DayViewViewModel extends BaseDbViewModel {

    private DayViewLiveData liveData = new DayViewLiveData();
    public LiveData<List<EventsItem>> events = liveData;
    @Nullable
    private EventsPagerItem item;

    public DayViewViewModel(Application application) {
        super(application);
    }

    public void setItem(@Nullable EventsPagerItem item) {
        this.item = item;
        liveData.update();
    }

    public void deleteBirthday(@NonNull Birthday birthday) {
        isInProgress.postValue(true);
        run(() -> {
            getAppDb().birthdaysDao().delete(birthday);
            end(() -> {
                isInProgress.postValue(false);
                result.postValue(Commands.DELETED);
                liveData.update();
            });
            new DeleteBirthdayFilesAsync(getApplication()).execute(birthday.getUuId());
        });
    }

    public void moveToTrash(@NonNull Reminder reminder) {
        isInProgress.postValue(true);
        run(() -> {
            reminder.setRemoved(true);
            EventControlFactory.getController(reminder).stop();
            getAppDb().reminderDao().insert(reminder);
            end(() -> {
                isInProgress.postValue(false);
                result.postValue(Commands.DELETED);
                Toast.makeText(getApplication(), R.string.deleted, Toast.LENGTH_SHORT).show();
                liveData.update();
            });
            new UpdateFilesAsync(getApplication()).execute(reminder);
        });
    }

    private class DayViewLiveData extends LiveData<List<EventsItem>> implements DayViewProvider.Callback, DayViewProvider.InitCallback {

        DayViewLiveData() {
            DayViewProvider provider = EventsDataSingleton.getInstance().getProvider();
            if (provider != null) {
                provider.addObserver(this);
            }
        }

        void update() {
            DayViewProvider provider = EventsDataSingleton.getInstance().getProvider();
            if (provider != null && item != null) {
                provider.findMatches(item.getDay(), item.getMonth(), item.getYear(), true, this);
            }
        }

        @Override
        public void apply(@NonNull List<EventsItem> list) {
            postValue(list);
        }

        @Override
        protected void onInactive() {
            super.onInactive();
            DayViewProvider provider = EventsDataSingleton.getInstance().getProvider();
            if (provider != null) {
                provider.removeCallback(this);
                provider.removeObserver(this);
            }
        }

        @Override
        public void onFinish() {
            update();
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/view_models/day_view/DayViewViewModel.kt
package com.elementary.tasks.core.view_models.day_view

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.elementary.tasks.birthdays.work.BirthdayDeleteBackupWorker
import com.elementary.tasks.core.controller.EventControlFactory
import com.elementary.tasks.core.data.AppDb
import com.elementary.tasks.core.data.models.Birthday
import com.elementary.tasks.core.data.models.Reminder
import com.elementary.tasks.core.data.models.ReminderGroup
import com.elementary.tasks.core.utils.Constants
import com.elementary.tasks.core.utils.Prefs
import com.elementary.tasks.core.utils.launchDefault
import com.elementary.tasks.core.utils.withUIContext
import com.elementary.tasks.core.view_models.BaseDbViewModel
import com.elementary.tasks.core.view_models.Commands
import com.elementary.tasks.day_view.DayViewProvider
import com.elementary.tasks.day_view.EventsPagerItem
import com.elementary.tasks.day_view.day.EventModel
import com.elementary.tasks.reminder.work.ReminderSingleBackupWorker
import kotlinx.coroutines.Job
import timber.log.Timber
import java.util.*

class DayViewViewModel(
  private val calculateFuture: Boolean,
  private val eventControlFactory: EventControlFactory,
  dayViewProvider: DayViewProvider,
  appDb: AppDb,
  prefs: Prefs
) : BaseDbViewModel(appDb, prefs) {

  private var _events: MutableLiveData<Pair<EventsPagerItem, List<EventModel>>> = MutableLiveData()
  var events: LiveData<Pair<EventsPagerItem, List<EventModel>>> = _events

  private var _groups: MutableList<ReminderGroup> = mutableListOf()
  val groups: List<ReminderGroup>
    get() = _groups

  private val liveData = DayViewLiveData(dayViewProvider)

  init {
    appDb.reminderGroupDao().loadAll().observeForever {
      if (it != null) {
        _groups.clear()
        _groups.addAll(it)
      }
    }
  }

  fun findEvents(item: EventsPagerItem) {
    try {
      liveData.findEvents(item, true) { eventsPagerItem, list ->
        _events.postValue(Pair(eventsPagerItem, list))
      }
    } catch (e: UninitializedPropertyAccessException) {
    }
  }

  fun saveReminder(reminder: Reminder) {
    postInProgress(true)
    launchDefault {
      appDb.reminderDao().insert(reminder)
      postInProgress(false)
      postCommand(Commands.SAVED)
      startWork(ReminderSingleBackupWorker::class.java, Constants.INTENT_ID, reminder.uuId)
    }
  }

  fun deleteBirthday(id: String) {
    postInProgress(true)
    launchDefault {
      appDb.birthdaysDao().delete(id)
      postInProgress(false)
      postCommand(Commands.DELETED)
      startWork(BirthdayDeleteBackupWorker::class.java, Constants.INTENT_ID, id)
    }
  }

  fun moveToTrash(reminder: Reminder) {
    postInProgress(true)
    launchDefault {
      val fromDb = appDb.reminderDao().getById(reminder.uuId)
      if (fromDb != null) {
        fromDb.isRemoved = true
        eventControlFactory.getController(fromDb).stop()
        appDb.reminderDao().insert(fromDb)
      }
      postInProgress(false)
      postCommand(Commands.DELETED)
      startWork(ReminderSingleBackupWorker::class.java, Constants.INTENT_ID, reminder.uuId)
    }
  }

  fun skip(reminder: Reminder) {
    postInProgress(true)
    launchDefault {
      val fromDb = appDb.reminderDao().getById(reminder.uuId)
      if (fromDb != null) {
        eventControlFactory.getController(fromDb).skip()
      }
      postInProgress(false)
      postCommand(Commands.DELETED)
      startWork(ReminderSingleBackupWorker::class.java, Constants.INTENT_ID, reminder.uuId)
    }
  }

  private inner class DayViewLiveData(
    private val dayViewProvider: DayViewProvider
  ) : LiveData<Pair<EventsPagerItem, List<EventModel>>>() {

    private val reminderData = ArrayList<EventModel>()
    private val birthdayData = ArrayList<EventModel>()
    private val birthdays = appDb.birthdaysDao().loadAll()
    private val reminders = appDb.reminderDao().loadType(active = true, removed = false)

    private var eventsPagerItem: EventsPagerItem? = null
    private var job: Job? = null
    private var listener: ((EventsPagerItem, List<EventModel>) -> Unit)? = null
    private var sort = false

    private val birthdayObserver: Observer<in List<Birthday>> = Observer { list ->
      Timber.d("birthdaysChanged: ")
      launchDefault {
        birthdayData.clear()
        birthdayData.addAll(
          list.map { dayViewProvider.toEventModel(it) }
        )
        repeatSearch()
      }
    }
    private val reminderObserver: Observer<in List<Reminder>> = Observer {
      Timber.d("remindersChanged: ")
      launchDefault {
        if (it != null) {
          reminderData.clear()
          reminderData.addAll(dayViewProvider.loadReminders(calculateFuture, it))
          repeatSearch()
        }
      }
    }

    init {
      birthdays.observeForever(birthdayObserver)
      reminders.observeForever(reminderObserver)
    }

    fun findEvents(eventsPagerItem: EventsPagerItem, sort: Boolean, listener: ((EventsPagerItem, List<EventModel>) -> Unit)?) {
      if (listener == null) return
      this.listener = listener
      this.eventsPagerItem = eventsPagerItem
      this.sort = sort
      val toSearch = mutableListOf<EventModel>()
      toSearch.addAll(birthdayData)
      toSearch.addAll(reminderData)
      findMatches(toSearch, eventsPagerItem, sort)
    }

    override fun onInactive() {
      super.onInactive()
      Timber.d("onInactive: ")
      birthdays.observeForever(birthdayObserver)
      reminders.observeForever(reminderObserver)
      this.eventsPagerItem = null
    }

    override fun onActive() {
      super.onActive()
      Timber.d("onActive: ")
      birthdays.removeObserver(birthdayObserver)
      reminders.removeObserver(reminderObserver)
    }

    private fun notifyObserver(eventsPagerItem: EventsPagerItem, list: List<EventModel>) {
      listener?.invoke(eventsPagerItem, list)
    }

    private fun repeatSearch() {
      val item = eventsPagerItem ?: return
      findEvents(item, this.sort, listener)
    }

    private fun findMatches(list: List<EventModel>, eventsPagerItem: EventsPagerItem, sort: Boolean) {
      this.job?.cancel()
      this.job = launchDefault {
        val res = ArrayList<EventModel>()
        Timber.d("Search events: $eventsPagerItem")
        for (item in list) {
          val mDay = item.day
          val mMonth = item.month
          val mYear = item.year
          val type = item.viewType
          if (type == EventModel.BIRTHDAY && mDay == eventsPagerItem.day && mMonth == eventsPagerItem.month) {
            res.add(item)
          } else {
            if (mDay == eventsPagerItem.day && mMonth == eventsPagerItem.month && mYear == eventsPagerItem.year) {
              res.add(item)
            }
          }
        }
        Timber.d("Search events: found -> %d", res.size)
        if (!sort) {
          withUIContext { notifyObserver(eventsPagerItem, res) }
        } else {
          val sorted = try {
            res.asSequence().sortedBy { it.getMillis() }.toList()
          } catch (e: IllegalArgumentException) {
            res
          }
          withUIContext { notifyObserver(eventsPagerItem, sorted) }
        }
      }
    }
  }
}
